x/programs/cmd/simulator/src/simulator.rs

```diff
@@ -0,0 +1,367 @@
+use crate::codec::{base64_decode, id_from_usize};
+use crate::util::get_path;
+use crate::{param::Param, Id};
+use borsh::BorshDeserialize;
+use serde::{Deserialize, Serialize};
+use std::{
+    io::{BufRead, BufReader, Write},
+    path::Path,
+    process::{Child, Command, Stdio},
+};
+use thiserror::Error;
+use wasmlanche_sdk::ExternalCallError;
+
+/// The endpoint to call for a [`SimulatorRequest`].
+#[derive(Debug, Serialize, PartialEq, Clone)]
+#[serde(rename_all = "lowercase")]
+pub enum Endpoint {
+    /// Make a read-only call to a program function and return the result.
+    ReadOnly,
+    /// Create a transaction on-chain from a possible state changing program
+    /// function call. A program's function can internally optionally call other
+    /// functions including program to program.
+    Execute,
+    /// Create a new program on-chain
+    CreateProgram,
+}
+
+#[derive(Error, Debug)]
+pub enum ClientError {
+    #[error("Read error: {0}")]
+    Read(#[from] std::io::Error),
+    #[error("EOF")]
+    Eof,
+    #[error("Missing handle")]
+    StdIo,
+}
+
+pub struct Simulator<W, R> {
+    writer: W,
+    responses: R,
+}
+
+pub fn build_simulator(
+) -> Result<Simulator<impl Write, impl Iterator<Item = SimulatorResponseItem>>, ClientError> {
+    let path = get_path();
```

Comment:
You can bring this into this file, and I would rename it to `get_binary_path` or something along those lines.